### PR TITLE
PLATFORM-1239: improve caching of DataProvider::GetTopFiveUser

### PR DESCRIPTION
**Pull request by macbre**
https://wikia-inc.atlassian.net/browse/PLATFORM-1239
- use `WikiaDataAccess::cacheWithLock`
- set the maximum limit value and cache all results + slice them before returning
- stop when all required results are collected
- code cleanup

@michalroszka / @wladekb / @harnash

### CODE REVIEW

File: extensions/wikia/DataProvider/DataProvider.php [hunk 1]
```diff
@@ -427,47 +429,78 @@ public static function GetNewlyCreatedArticles($limit = 5, $ns = array(NS_MAIN))
...
+		$limit = min( $limit, self::TOP_USERS_MAX_LIMIT );
 
 		if (empty($wgStatsDB) || empty($wgStatsDBEnabled)) {
-			$result = array();
+			$results = [];
 		}
 		else {
-			$memckey = wfMemcKey("TopFiveUsers", $limit);
-			$results = $wgMemc->get($memckey);
+			$results = WikiaDataAccess::cacheWithLock( wfMemcKey( __CLASS__, 'GetTopFiveUsers' ), WikiaResponse::CACHE_STANDARD, function() use ($fname) {
```
Comment (wladekb) [inline extensions/wikia/DataProvider/DataProvider.php]: AFAIR we used to put __METHOD__ into wfMemcKey in such cases

File: extensions/wikia/DataProvider/DataProvider.php [hunk 2]
```diff
@@ -427,47 +429,78 @@ public static function GetNewlyCreatedArticles($limit = 5, $ns = array(NS_MAIN))
...
+					"user_groups",
+					"ug_user",
+					[ 'ug_group' => [ 'staff', 'bot' ] ],
+					$fname
+				);
+
+				// add more blacklisted user IDs
+				$users_list[] = 0;
+				$users_list[] = 22439; // Wikia
+				$users_list[] = 929702; // CreateWiki script
```
Comment (wladekb) [inline extensions/wikia/DataProvider/DataProvider.php]: Good candidate to put the list of such IDs into single common place like config variable and use it here

File: extensions/wikia/DataProvider/DataProvider.php (hunk 1 above)
Comment (macbre) [inline extensions/wikia/DataProvider/DataProvider.php]: We do, but this call is within a closure.

File: extensions/wikia/DataProvider/DataProvider.php [hunk 3]
```diff
@@ -427,47 +429,78 @@ public static function GetNewlyCreatedArticles($limit = 5, $ns = array(NS_MAIN))
...
-
-				$res = $dbs->query($dbs->limitResult($query, $limit * 4, 0));
-
-				$results = array();
+				$users_list = $dbr->selectFieldValues(
+					"user_groups",
+					"ug_user",
+					[ 'ug_group' => [ 'staff', 'bot' ] ],
+					$fname
+				);
```
Comment (wladekb) [inline extensions/wikia/DataProvider/DataProvider.php]: use DISTINCT here?

File: extensions/wikia/DataProvider/DataProvider.php (hunk 1 above)
Comment (macbre) [inline extensions/wikia/DataProvider/DataProvider.php]: Isn't it a bit easier to grep this key name when it's provided as a raw text?

File: extensions/wikia/DataProvider/DataProvider.php (hunk 2 above)
Comment (macbre) [inline extensions/wikia/DataProvider/DataProvider.php]: Good point, but I'm afraid that we will end up with completely rewritten code here :)

File: extensions/wikia/DataProvider/DataProvider.php (hunk 3 above)
Comment (macbre) [inline extensions/wikia/DataProvider/DataProvider.php]: Improved in 0578e68

**Comment by michalroszka** [comment]
:+1:

**Pull request merged**